zkstack_cli/crates/zkstack/src/commands/dev/commands/v28_precompiles.rs

```diff
@@ -0,0 +1,303 @@
+use zkstack_cli_config::traits::{ReadConfig, ZkStackConfig};
+use zksync_basic_types::{
+    protocol_version::ProtocolVersionId, web3::Bytes, Address, L1BatchNumber, L2BlockNumber, U256,
+};
+use zksync_types::L2_BRIDGEHUB_ADDRESS;
+use zksync_web3_decl::{
+    client::{DynClient, L2},
+    namespaces::ZksNamespaceClient,
+};
+
+use crate::{
+    abi::{BridgehubAbi, ZkChainAbi},
+    commands::{
+        chain::{
+            admin_call_builder::{AdminCall, AdminCallBuilder},
+            utils::get_default_foundry_path,
+        },
+        dev::commands::upgrade_utils::{print_error, set_upgrade_timestamp_calldata},
+    },
+};
+
+#[derive(Debug, Default)]
+pub struct FetchedChainInfo {
+    hyperchain_addr: Address,
+    chain_admin_addr: Address,
+    gw_hyperchain_addr: Address,
+    gw_chain_admin_addr: Address,
+    l1_asset_router_proxy: Address,
+    settlement_layer: u64,
+}
+
+async fn verify_next_batch_new_version(
+    batch_number: u32,
+    main_node_client: &DynClient<L2>,
+) -> anyhow::Result<()> {
+    let (_, right_bound) = main_node_client
+        .get_l2_block_range(L1BatchNumber(batch_number))
+        .await?
+        .context("Range must be present for a batch")?;
+
+    let next_l2_block = right_bound + 1;
+
+    let block_details = main_node_client
+        .get_block_details(L2BlockNumber(next_l2_block.as_u32()))
+        .await?
+        .with_context(|| format!("No L2 block is present after the batch {}", batch_number))?;
+
+    let protocol_version = block_details.protocol_version.with_context(|| {
+        format!(
+            "Protocol version not present for block {}",
+            next_l2_block.as_u64()
+        )
+    })?;
+    anyhow::ensure!(
+        protocol_version >= ProtocolVersionId::Version28,
+        "THe block does not yet contain the v28 (Precompiles) upgrade"
+    );
+
+    Ok(())
+}
+
+pub async fn check_chain_readiness(
+    l1_rpc_url: String,
+    l2_rpc_url: String,
+    gw_rpc_url: String,
+    l2_chain_id: u64,
+    gw_chain_id: u64,
+    settlement_layer: u64,
+) -> anyhow::Result<()> {
+    let l1_provider = get_ethers_provider(&l1_rpc_url)?;
+
+    let l2_client = get_zk_client(&l2_rpc_url, l2_chain_id)?;
+
+    let gw_client = get_ethers_provider(&gw_rpc_url)?;
+    let gw_client_l2 = get_zk_client(&gw_rpc_url, gw_chain_id)?;
+
+    if settlement_layer == gw_chain_id {
+        // GW
+        let diamond_proxy_addr = (BridgehubAbi::new(L2_BRIDGEHUB_ADDRESS, gw_client.clone()))
+            .get_zk_chain(l2_chain_id.into())
+            .await?;
+        let zkchain = ZkChainAbi::new(diamond_proxy_addr, gw_client.clone());
+        let batches_committed = zkchain.get_total_batches_committed().await?.as_u32();
+        let batches_verified = zkchain.get_total_batches_verified().await?.as_u32();
+
+        verify_next_batch_new_version(batches_committed, &gw_client_l2).await?;
+        verify_next_batch_new_version(batches_verified, &gw_client_l2).await?;
```

Comment:
where you are asking whther `batches_committed` batch on GW has the right version, you should check whether the chains' batch is correct (i.e. the same check as below), you can basically just move down and reuse this part:

```
        let batches_committed = zkchain.get_total_batches_committed().await?.as_u32();
        let batches_verified = zkchain.get_total_batches_verified().await?.as_u32();

        verify_next_batch_new_version(batches_committed, &l2_client).await?;
        verify_next_batch_new_version(batches_verified, &l2_client).await?;

```